scripts/packages/wix/make_msi_lib.ps1

```diff
@@ -0,0 +1,191 @@
+$Icon = "site\images\favicon.ico"
+$ArtworkBig = "scripts\packages\wix\dialog.bmp"
+$ArtworkSmall = "scripts\packages\wix\banner.bmp"
+$Licence = "scripts\packages\wix\license.rtf"
+$Guids = "scripts\packages\wix\guids.txt"
+$Wxs = "scripts\packages\wix\bazelmsi.wxs"
+
+# Logs a message to stdout.
+function Log-Info {
+    param([string]$msg)
+    $basename = $PSCommandPath.Substring($PSCommandPath.LastIndexOfAny("/\") + 1)
+    $time = (Get-Date).ToString("HH:mm:ss")
+    Write-Host "INFO[$basename $time] $msg"
+}
+
+function Replace-Slashes {
+    param([string]$s)
+    return $s.Replace("/", "\")
+}
+
+# Computes the release name from the 'BazelExe'.
+# 'BazelExe' must already have been validated to match this regex.
+# 'ReleaseName' is like "0.28.0rc5"
+# 'Version' is the SemVer part of 'ReleaseName', e.g. "0.28.0"
+function Compute-RelaseNameAndVersion {
+    param([Parameter(Mandatory=$true)][string]$BazelExe)
+
+    $rel = [regex]::Match($BazelExe, "bazel-([^-]+)-windows.*\.exe$").captures.groups[1].value
+    $ver = [regex]::Match($rel, "^([0-9]+\.[0-9]+\.[0-9]+)(rc[0-9]+)?$").captures.groups[1].value
+    return $rel, $ver
+}
+
+# Generates a new GUID, prints it as uppercase.
+# The WiX Toolkit expects uppercase GUIDs in the .wxs file.
+function Generate-Guid {
+    return [guid]::NewGuid().ToString().ToUpper()
+}
+
+# Returns the UpgradeGuid for this release.
+function Get-UpgradeGuid {
+    param([Parameter(Mandatory=$true)][string]$release_name)
+
+    $d = @{}
+    $result = $null
+    foreach ($line in Get-Content -Path $Guids) {
+        $line = "$line".Split('#')[0]
+        if ($line) {
+            $k, $v = $line.Split(' ', 2)
+            # Ensure all version prefixes in the file are unique.
+            if ($d.ContainsKey($k)) {
+                throw "Duplicate relase prefix $k in $Guids"
+            } else {
+                $d[$k] = $null
+            }
+            # Ensure all GUIDs in the file are unique. We can use the same
+            # hashtable because the value domains are distinct.
+            if ($d.ContainsKey($v)) {
+                throw "Duplicate GUID $v in $Guids"
+            } else {
+                $d[$v] = $null
+            }
+            if (! $result -and $release_name.StartsWith($k)) {
+                $result = "$v"
+                # Do not return yet, so we check that all GUIDs are unique.
+            }
+        }
+    }
+    if ($result) {
+        return "$result"
+    } else {
+        throw "UpgradeGuid for $release_name not found in $Guids"
+    }
+}
+
+# Returns the Bazel version (as a SemVer string) from the release name.
+function Get-BazelVersion {
+    if ("$ReleaseName" -match "rc[0-9]+$") {
+        return "$ReleaseName".Substring(0, "$ReleaseName".LastIndexOf("rc"))
+    } else {
+        return "$ReleaseName"
+    }
+}
+
+# Downloads and extracts the WiX Toolkit.
+# Returns the path where the tools are (e.g. "candle.exe").
+function Download-Wix {
```

Comment:
I'm not really comfortable with downloading Wix here.
This is fine for now, but I think we should vendor in tools like this, especially when we get to code signing.
And that brings up the whole question of if we should continue to bundle all our dependencies in the distribution tarball (no) or call them out by reference (yes).

Reply:
Thanks for the observation.

I think downloading is simpler than vendoring, and unlike vendoring it raises no questions about licencing. So I'd like to keep it like so for now.